consensus/istanbul/backend/announce.go

```diff
@@ -17,405 +17,671 @@
+
+	// Create all the variables needed for the periodic gossip
+	var announceGossipTicker *time.Ticker
+	var announceGossipTickerCh <-chan time.Time
+	var announceGossipFrequencyState AnnounceGossipFrequencyState
+	var currentAnnounceGossipTickerDuration time.Duration
+	var numGossipedMsgsInHighFreqAfterFirstPeerState int
+
+	for {
+		select {
+		case <-checkIfShouldAnnounceTicker.C:
+			logger.Trace("Checking if this node should announce it's enode")
+
+			shouldAnnounce, err := sb.shouldGenerateAndProcessAnnounce()
+			if err != nil {
+				logger.Warn("Error in checking if should announce", err)
+				break
+			}
+
+			if shouldAnnounce && announceGossipTickerCh == nil {
+				// Immediately gossip an announce
+				go sb.gossipAnnounce()
+
+				if sb.config.AnnounceAggressiveGossipOnEnablement {
+					announceGossipFrequencyState = HighFreqBeforeFirstPeerState
 
-func (ar *announceRecord) String() string {
-	return fmt.Sprintf("{DestAddress: %s, EncryptedEnodeURL length: %d}", ar.DestAddress.String(), len(ar.EncryptedEnodeURL))
+					// Send an announce message once a minute
+					currentAnnounceGossipTickerDuration = 1 * time.Minute
+
+					numGossipedMsgsInHighFreqAfterFirstPeerState = 0
+				} else {
+					announceGossipFrequencyState = LowFreqState
+					currentAnnounceGossipTickerDuration = time.Duration(sb.config.AnnounceGossipPeriod) * time.Second
+				}
+
+				// Enable periodic gossiping by setting announceGossipTickerCh to non nil value
+				announceGossipTicker = time.NewTicker(currentAnnounceGossipTickerDuration)
+				announceGossipTickerCh = announceGossipTicker.C
+				logger.Trace("Enabled periodic gossiping of announce message")
+			} else if !shouldAnnounce && announceGossipTickerCh != nil {
+				// Disable periodic gossiping by setting announceGossipTickerCh to nil
+				announceGossipTicker.Stop()
+				announceGossipTickerCh = nil
+				logger.Trace("Disabled periodic gossiping of announce message")
+			}
+
+			// Use this timer to also prune all announce related data structures.
+			if err := sb.pruneAnnounceDataStructures(); err != nil {
+				logger.Warn("Error is pruning announce data structures", "err", err)
+			}
+
+		case <-announceGossipTickerCh: // If this is nil (when shouldAnnounce was most recently false), this channel will never receive an event
+			logger.Trace("Going to gossip an announce message", "announceGossipFrequencyState", announceGossipFrequencyState, "numGossipedMsgsInHighFreqAfterFirstPeerState", numGossipedMsgsInHighFreqAfterFirstPeerState)
+			switch announceGossipFrequencyState {
+			case HighFreqBeforeFirstPeerState:
+				if len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) > 0 {
+					announceGossipFrequencyState = HighFreqAfterFirstPeerState
+				}
+
+			case HighFreqAfterFirstPeerState:
+				if numGossipedMsgsInHighFreqAfterFirstPeerState >= 10 {
+					announceGossipFrequencyState = LowFreqState
+				}
+				numGossipedMsgsInHighFreqAfterFirstPeerState += 1
+
+			case LowFreqState:
```

Comment:
How this is implemented right now results in a slightly odd edge case in that 12 announce messages will be sent at high frequency before actually slowing down, when it seems the intention may be for 10. +1 because it checks `numGossipedMsgsInHighFreqAfterFirstPeerState` before incrementing it, and +1 because the ticker remains at high frequency for one more iteration. It's also a little odd in that the loop invariant of `currentAnnounceGossipTickerDuration` matching `announceGossipFrequencyState` is broken for one iteration. Would it make sense to move the code from the `LowFreqState` into the if statement of the one above?

Reply:
do you have a strong opinion about this code block?  This code is mostly mean for backward's compatibility between phase 2 and phase 3 nodes, and the plan is to remove this altogether for phase 4.

Reply:
I didn't really think about that. I wouldn't worry too much about it perfecting it then.